src/appleseed/foundation/image/genericimagefilewriter.cpp

```diff
@@ -31,48 +30,488 @@
 #include "genericimagefilewriter.h"
 
 // appleseed.foundation headers.
-#include "foundation/core/exceptions/exceptionunsupportedfileformat.h"
-#include "foundation/image/exrimagefilewriter.h"
-#include "foundation/image/pngimagefilewriter.h"
+#include "foundation/core/exceptions/exceptionioerror.h"
+#include "foundation/image/icanvas.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/compiler.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/iostreamop.h"
 #include "foundation/utility/string.h"
 
 // Boost headers.
 #include "boost/filesystem/path.hpp"
 
 // Standard headers.
-#include <string>
-
-using namespace std;
-namespace bf = boost::filesystem;
+#include <algorithm>
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <vector>
 
 namespace foundation
 {
 
-//
-// GenericImageFileWriter class implementation.
-//
+struct GenericImageFileWriter::Impl
+{
+    std::vector<const ICanvas*>     m_canvas;
+    std::vector<OIIO::ImageSpec>    m_spec;
+};
+
+GenericImageFileWriter::GenericImageFileWriter(const char* filename) :
+    IImageFileWriter(),
+    impl(new Impl())
+{
+    assert(filename);
+
+    m_filename = filename;
+
+    m_writer = OIIO::ImageOutput::create(m_filename);
+    if (m_writer == nullptr)
+    {
+        const std::string msg = OIIO::geterror();
+        throw ExceptionIOError(msg.c_str());
+    }
+}
+
+GenericImageFileWriter::~GenericImageFileWriter()
+{
+    // Destroy the ImageOutput stucture.
+    OIIO::ImageOutput::destroy(m_writer);
```

Comment:
Is it legal to pass `nullptr` to `OIIO::ImageOutput::destroy()`? If not then you'll need to guard against this since `m_writer` can be null.